{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "444cd7c9-0b7c-4d33-9ae2-05d1d9d3e697",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from kipoiseq.extractors import MultiSampleVCF\n",
    "from tqdm import tqdm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "89803c94-9676-4940-b8fb-74c038d55fcd",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "\n",
    "try:\n",
    "    snakemake\n",
    "except NameError:\n",
    "    from snakemk_util import load_rule_args\n",
    "    snakemake = load_rule_args(\n",
    "        snakefile = os.getcwd() + '/../Snakefile',\n",
    "        rule_name = 'prioritize_vep_variants',\n",
    "        root=os.getcwd() + \"/..\",\n",
    "        default_wildcards={'chrom': 'chr10'}\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "89344d9f-7d50-4447-803e-7f2a31db148a",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_metadata = pd.read_csv(snakemake.input['released_files'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "3ba90070-64ad-413b-b305-c87065b9e297",
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_mapping = {\n",
    "    row['CGND_ID']: row['Participant_ID']\n",
    "    for i, row in df_metadata[df_metadata['omic'] == 'genomics'].iterrows()\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "c72b66ee-094b-4190-9ffe-59afd846cd4f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sampleID</th>\n",
       "      <th>pValue</th>\n",
       "      <th>padjust</th>\n",
       "      <th>zScore</th>\n",
       "      <th>l2fc</th>\n",
       "      <th>rawcounts</th>\n",
       "      <th>normcounts</th>\n",
       "      <th>meanCorrected</th>\n",
       "      <th>theta</th>\n",
       "      <th>aberrant</th>\n",
       "      <th>AberrantBySample</th>\n",
       "      <th>AberrantByGene</th>\n",
       "      <th>padj_rank</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_biotype</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>geneID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000000003</th>\n",
       "      <td>CASE.NEUAA599TMX</td>\n",
       "      <td>1.036564e-17</td>\n",
       "      <td>9.877520e-13</td>\n",
       "      <td>-9.49</td>\n",
       "      <td>-4.48</td>\n",
       "      <td>30</td>\n",
       "      <td>40.89</td>\n",
       "      <td>936.79</td>\n",
       "      <td>19.46</td>\n",
       "      <td>True</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>2.0</td>\n",
       "      <td>X</td>\n",
       "      <td>100627108.0</td>\n",
       "      <td>100639991.0</td>\n",
       "      <td>-</td>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000003</th>\n",
       "      <td>CTRL.NEUHE723FGT</td>\n",
       "      <td>6.894534e-09</td>\n",
       "      <td>1.313973e-03</td>\n",
       "      <td>-5.45</td>\n",
       "      <td>-2.59</td>\n",
       "      <td>77</td>\n",
       "      <td>154.45</td>\n",
       "      <td>936.79</td>\n",
       "      <td>19.46</td>\n",
       "      <td>True</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>X</td>\n",
       "      <td>100627108.0</td>\n",
       "      <td>100639991.0</td>\n",
       "      <td>-</td>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000003</th>\n",
       "      <td>CASE.NEUBA169GXD</td>\n",
       "      <td>3.543101e-08</td>\n",
       "      <td>1.688127e-03</td>\n",
       "      <td>-5.01</td>\n",
       "      <td>-2.39</td>\n",
       "      <td>140</td>\n",
       "      <td>179.07</td>\n",
       "      <td>936.79</td>\n",
       "      <td>19.46</td>\n",
       "      <td>True</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>4.0</td>\n",
       "      <td>X</td>\n",
       "      <td>100627108.0</td>\n",
       "      <td>100639991.0</td>\n",
       "      <td>-</td>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000003056</th>\n",
       "      <td>CTRL.NEUCV136DHM</td>\n",
       "      <td>4.702285e-13</td>\n",
       "      <td>4.480852e-09</td>\n",
       "      <td>6.78</td>\n",
       "      <td>0.50</td>\n",
       "      <td>3296</td>\n",
       "      <td>2658.65</td>\n",
       "      <td>1884.46</td>\n",
       "      <td>462.00</td>\n",
       "      <td>True</td>\n",
       "      <td>102</td>\n",
       "      <td>1</td>\n",
       "      <td>20.0</td>\n",
       "      <td>12</td>\n",
       "      <td>8940362.0</td>\n",
       "      <td>8949955.0</td>\n",
       "      <td>-</td>\n",
       "      <td>M6PR</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000004455</th>\n",
       "      <td>CASE.NEUGW340YEB</td>\n",
       "      <td>8.518985e-12</td>\n",
       "      <td>1.623564e-06</td>\n",
       "      <td>-7.12</td>\n",
       "      <td>-0.75</td>\n",
       "      <td>942</td>\n",
       "      <td>816.87</td>\n",
       "      <td>1374.93</td>\n",
       "      <td>241.90</td>\n",
       "      <td>True</td>\n",
       "      <td>12</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>33007939.0</td>\n",
       "      <td>33080996.0</td>\n",
       "      <td>-</td>\n",
       "      <td>AK2</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000278771</th>\n",
       "      <td>CASE.NEUGW340YEB</td>\n",
       "      <td>3.340853e-11</td>\n",
       "      <td>3.183531e-06</td>\n",
       "      <td>5.25</td>\n",
       "      <td>2.22</td>\n",
       "      <td>4043</td>\n",
       "      <td>2203.72</td>\n",
       "      <td>471.14</td>\n",
       "      <td>10.14</td>\n",
       "      <td>True</td>\n",
       "      <td>12</td>\n",
       "      <td>1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>14</td>\n",
       "      <td>49853615.0</td>\n",
       "      <td>49853914.0</td>\n",
       "      <td>-</td>\n",
       "      <td>Metazoa_SRP</td>\n",
       "      <td>misc_RNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000278777</th>\n",
       "      <td>CASE.NEUZV656DD1</td>\n",
       "      <td>2.250425e-06</td>\n",
       "      <td>2.859269e-02</td>\n",
       "      <td>4.67</td>\n",
       "      <td>2.44</td>\n",
       "      <td>6</td>\n",
       "      <td>8.22</td>\n",
       "      <td>0.45</td>\n",
       "      <td>22.53</td>\n",
       "      <td>True</td>\n",
       "      <td>17</td>\n",
       "      <td>1</td>\n",
       "      <td>15.0</td>\n",
       "      <td>GL000220.1</td>\n",
       "      <td>108279.0</td>\n",
       "      <td>108340.0</td>\n",
       "      <td>+</td>\n",
       "      <td>AL592188.12</td>\n",
       "      <td>miRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000279203</th>\n",
       "      <td>CASE.NEUZP278MR4</td>\n",
       "      <td>1.152811e-08</td>\n",
       "      <td>8.788194e-05</td>\n",
       "      <td>4.55</td>\n",
       "      <td>1.10</td>\n",
       "      <td>268</td>\n",
       "      <td>124.23</td>\n",
       "      <td>57.45</td>\n",
       "      <td>56.39</td>\n",
       "      <td>True</td>\n",
       "      <td>32</td>\n",
       "      <td>1</td>\n",
       "      <td>25.0</td>\n",
       "      <td>19</td>\n",
       "      <td>15383201.0</td>\n",
       "      <td>15383832.0</td>\n",
       "      <td>-</td>\n",
       "      <td>AC005785.5</td>\n",
       "      <td>TEC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000279500</th>\n",
       "      <td>CASE.NEUMD869UV7</td>\n",
       "      <td>6.831112e-08</td>\n",
       "      <td>3.254716e-03</td>\n",
       "      <td>-7.31</td>\n",
       "      <td>-6.41</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>94.13</td>\n",
       "      <td>6.55</td>\n",
       "      <td>True</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>4.0</td>\n",
       "      <td>12</td>\n",
       "      <td>128813185.0</td>\n",
       "      <td>128814750.0</td>\n",
       "      <td>-</td>\n",
       "      <td>RP11-21K12.2</td>\n",
       "      <td>TEC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000279598</th>\n",
       "      <td>CASE.NEUJA217MTJ</td>\n",
       "      <td>1.124608e-14</td>\n",
       "      <td>7.144332e-10</td>\n",
       "      <td>-9.37</td>\n",
       "      <td>-3.85</td>\n",
       "      <td>14</td>\n",
       "      <td>10.93</td>\n",
       "      <td>169.03</td>\n",
       "      <td>23.17</td>\n",
       "      <td>True</td>\n",
       "      <td>63</td>\n",
       "      <td>1</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2</td>\n",
       "      <td>178554560.0</td>\n",
       "      <td>178558966.0</td>\n",
       "      <td>+</td>\n",
       "      <td>RP11-65L3.3</td>\n",
       "      <td>TEC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1396 rows × 19 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                         sampleID        pValue       padjust  zScore  l2fc  \\\n",
       "geneID                                                                        \n",
       "ENSG00000000003  CASE.NEUAA599TMX  1.036564e-17  9.877520e-13   -9.49 -4.48   \n",
       "ENSG00000000003  CTRL.NEUHE723FGT  6.894534e-09  1.313973e-03   -5.45 -2.59   \n",
       "ENSG00000000003  CASE.NEUBA169GXD  3.543101e-08  1.688127e-03   -5.01 -2.39   \n",
       "ENSG00000003056  CTRL.NEUCV136DHM  4.702285e-13  4.480852e-09    6.78  0.50   \n",
       "ENSG00000004455  CASE.NEUGW340YEB  8.518985e-12  1.623564e-06   -7.12 -0.75   \n",
       "...                           ...           ...           ...     ...   ...   \n",
       "ENSG00000278771  CASE.NEUGW340YEB  3.340853e-11  3.183531e-06    5.25  2.22   \n",
       "ENSG00000278777  CASE.NEUZV656DD1  2.250425e-06  2.859269e-02    4.67  2.44   \n",
       "ENSG00000279203  CASE.NEUZP278MR4  1.152811e-08  8.788194e-05    4.55  1.10   \n",
       "ENSG00000279500  CASE.NEUMD869UV7  6.831112e-08  3.254716e-03   -7.31 -6.41   \n",
       "ENSG00000279598  CASE.NEUJA217MTJ  1.124608e-14  7.144332e-10   -9.37 -3.85   \n",
       "\n",
       "                 rawcounts  normcounts  meanCorrected   theta  aberrant  \\\n",
       "geneID                                                                    \n",
       "ENSG00000000003         30       40.89         936.79   19.46      True   \n",
       "ENSG00000000003         77      154.45         936.79   19.46      True   \n",
       "ENSG00000000003        140      179.07         936.79   19.46      True   \n",
       "ENSG00000003056       3296     2658.65        1884.46  462.00      True   \n",
       "ENSG00000004455        942      816.87        1374.93  241.90      True   \n",
       "...                    ...         ...            ...     ...       ...   \n",
       "ENSG00000278771       4043     2203.72         471.14   10.14      True   \n",
       "ENSG00000278777          6        8.22           0.45   22.53      True   \n",
       "ENSG00000279203        268      124.23          57.45   56.39      True   \n",
       "ENSG00000279500          0        0.00          94.13    6.55      True   \n",
       "ENSG00000279598         14       10.93         169.03   23.17      True   \n",
       "\n",
       "                 AberrantBySample  AberrantByGene  padj_rank  Chromosome  \\\n",
       "geneID                                                                     \n",
       "ENSG00000000003                 5               3        2.0           X   \n",
       "ENSG00000000003                 4               3        1.0           X   \n",
       "ENSG00000000003                 4               3        4.0           X   \n",
       "ENSG00000003056               102               1       20.0          12   \n",
       "ENSG00000004455                12               1        1.0           1   \n",
       "...                           ...             ...        ...         ...   \n",
       "ENSG00000278771                12               1        2.0          14   \n",
       "ENSG00000278777                17               1       15.0  GL000220.1   \n",
       "ENSG00000279203                32               1       25.0          19   \n",
       "ENSG00000279500                 5               1        4.0          12   \n",
       "ENSG00000279598                63               1        3.0           2   \n",
       "\n",
       "                       Start          End Strand     gene_name    gene_biotype  \n",
       "geneID                                                                          \n",
       "ENSG00000000003  100627108.0  100639991.0      -        TSPAN6  protein_coding  \n",
       "ENSG00000000003  100627108.0  100639991.0      -        TSPAN6  protein_coding  \n",
       "ENSG00000000003  100627108.0  100639991.0      -        TSPAN6  protein_coding  \n",
       "ENSG00000003056    8940362.0    8949955.0      -          M6PR  protein_coding  \n",
       "ENSG00000004455   33007939.0   33080996.0      -           AK2  protein_coding  \n",
       "...                      ...          ...    ...           ...             ...  \n",
       "ENSG00000278771   49853615.0   49853914.0      -   Metazoa_SRP        misc_RNA  \n",
       "ENSG00000278777     108279.0     108340.0      +   AL592188.12           miRNA  \n",
       "ENSG00000279203   15383201.0   15383832.0      -    AC005785.5             TEC  \n",
       "ENSG00000279500  128813185.0  128814750.0      -  RP11-21K12.2             TEC  \n",
       "ENSG00000279598  178554560.0  178558966.0      +   RP11-65L3.3             TEC  \n",
       "\n",
       "[1396 rows x 19 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_results = pd.read_csv(snakemake.input['results']).set_index('geneID')\n",
    "df_genes = pd.read_csv(snakemake.input['genes']).set_index('geneID')\n",
    "df_results = df_results.join(df_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "40c52eef-886d-4e9c-b444-dcc7c762d38a",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_results = df_results[df_results['Chromosome'] == snakemake.wildcards['chrom'].replace('chr', '')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "c928d3f1-f7d2-4f50-b607-26cbe63f84d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_parquet(snakemake.input['vep'])\n",
    "df['variant'] = snakemake.wildcards['chrom'] + ':' + df['pos'].astype(str) + ':'+ df['ref'] + '>' + df['alt']\n",
    "df = df.set_index('variant')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "0e8790ff-5907-4077-a819-05cc2a546510",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_mmsplice = pd.read_csv(snakemake.input['mmsplice'])\n",
    "\n",
    "df_mmsplice = df_mmsplice[df_mmsplice['delta_logit_psi'].abs() > 2] \\\n",
    "    .rename(columns={'gene_id': 'geneID', 'ID': 'variant'})\n",
    "\n",
    "df_mmsplice['geneID'] = df_mmsplice['geneID'].str.split('.').str.get(0)\n",
    "df_mmsplice = df_mmsplice[['geneID', 'variant', 'delta_logit_psi']].drop_duplicates()\n",
    "\n",
    "df_mmsplice = df_mmsplice[df_mmsplice['variant'].str.startswith(snakemake.wildcards['chrom'])]\n",
    "df_mmsplice = df_mmsplice.groupby(['geneID', 'variant']).agg(max).reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "a464e441-cc3e-4fc7-9782-6f5ea3a653d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df.join(df_mmsplice.set_index('variant')[['delta_logit_psi']])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "46da1937-5266-4c91-ae09-43ef640b3469",
   "metadata": {},
   "outputs": [],
   "source": [
    "# if high scored mmsplice moderate \n",
    "df.loc[(~df['delta_logit_psi'].isna()) & (df['IMPACT'] != 'HIGH'), 'IMPACT'] = 'MODERATE'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "cf389252-8ee8-4a52-b04c-6e808227086b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(2014821, 90)\n",
      "(6656, 90)\n",
      "(202, 90)\n"
     ]
    }
   ],
   "source": [
    "df = df[df['AF'].fillna(0) < 0.001] # subset outliers\n",
    "print(df.shape)\n",
    "df = df[df['IMPACT'].isin({'HIGH', 'MODERATE'})] # subset for high impact variants\n",
    "print(df.shape)\n",
    "df = df[df['Gene'].isin(set(df_results.index))] # subset outliers\n",
    "print(df.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "5c252069-69cb-4ec0-9fcd-2d9c57021da0",
   "metadata": {},
   "outputs": [],
   "source": [
    "vcf = MultiSampleVCF(snakemake.input['vcf'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "fc970c22-c89b-4434-97b9-2505cfb1e8c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_mmsplice['samples'] = [\n",
    "    [\n",
    "        sample_mapping[k.replace('-b38', '')]\n",
    "        for k, v in vcf.get_samples(vcf.get_variant(i)).items()\n",
    "    ]\n",
    "    for i in df_mmsplice['variant']\n",
    "]\n",
    "df_mmsplice = df_mmsplice[df_mmsplice['samples'].map(len) < 7].explode('samples')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "add61c45-5951-460f-85e4-a75ed8d28fb6",
   "metadata": {},
   "outputs": [],
   "source": [
    "df['samples'] = [\n",
    "    [\n",
    "        sample_mapping[k.replace('-b38', '')]\n",
    "        for k, v in vcf.get_samples(vcf.get_variant(i)).items()\n",
    "    ]\n",
    "    for i in df.index\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "417e80ff-bb01-4da1-bef3-de62e74f2bf7",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df[df['samples'].map(len) < 7].explode('samples')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "d62980cd-6d5c-4302-8c37-34f893e050f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_csv(snakemake.output['prioritized'])"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:anaconda-als]",
   "language": "python",
   "name": "conda-env-anaconda-als-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
