{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "444cd7c9-0b7c-4d33-9ae2-05d1d9d3e697",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from kipoiseq.extractors import MultiSampleVCF\n",
    "from tqdm import tqdm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "89803c94-9676-4940-b8fb-74c038d55fcd",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "\n",
    "try:\n",
    "    snakemake\n",
    "except NameError:\n",
    "    from snakemk_util import load_rule_args\n",
    "    snakemake = load_rule_args(\n",
    "        snakefile = os.getcwd() + '/../Snakefile',\n",
    "        rule_name = 'prioritize_vep_variants',\n",
    "        root=os.getcwd() + \"/..\",\n",
    "        default_wildcards={'chrom': 'chr10'}\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "89344d9f-7d50-4447-803e-7f2a31db148a",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_metadata = pd.read_csv(snakemake.input['released_files'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c72b66ee-094b-4190-9ffe-59afd846cd4f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sampleID</th>\n",
       "      <th>pValue</th>\n",
       "      <th>padjust</th>\n",
       "      <th>zScore</th>\n",
       "      <th>l2fc</th>\n",
       "      <th>rawcounts</th>\n",
       "      <th>normcounts</th>\n",
       "      <th>meanCorrected</th>\n",
       "      <th>theta</th>\n",
       "      <th>aberrant</th>\n",
       "      <th>AberrantBySample</th>\n",
       "      <th>AberrantByGene</th>\n",
       "      <th>padj_rank</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_biotype</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>geneID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000000003</th>\n",
       "      <td>CASE.NEUAA599TMX</td>\n",
       "      <td>7.156284e-18</td>\n",
       "      <td>1.363858e-12</td>\n",
       "      <td>-9.53</td>\n",
       "      <td>-4.51</td>\n",
       "      <td>30</td>\n",
       "      <td>40.08</td>\n",
       "      <td>934.62</td>\n",
       "      <td>19.47</td>\n",
       "      <td>True</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>X</td>\n",
       "      <td>100627108</td>\n",
       "      <td>100639991</td>\n",
       "      <td>-</td>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000003</th>\n",
       "      <td>CTRL.NEUHE723FGT</td>\n",
       "      <td>6.700160e-09</td>\n",
       "      <td>1.276929e-03</td>\n",
       "      <td>-5.45</td>\n",
       "      <td>-2.59</td>\n",
       "      <td>77</td>\n",
       "      <td>154.18</td>\n",
       "      <td>934.62</td>\n",
       "      <td>19.47</td>\n",
       "      <td>True</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>X</td>\n",
       "      <td>100627108</td>\n",
       "      <td>100639991</td>\n",
       "      <td>-</td>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000003</th>\n",
       "      <td>CASE.NEUBA169GXD</td>\n",
       "      <td>4.492763e-08</td>\n",
       "      <td>2.140598e-03</td>\n",
       "      <td>-4.95</td>\n",
       "      <td>-2.36</td>\n",
       "      <td>140</td>\n",
       "      <td>181.84</td>\n",
       "      <td>934.62</td>\n",
       "      <td>19.47</td>\n",
       "      <td>True</td>\n",
       "      <td>7</td>\n",
       "      <td>3</td>\n",
       "      <td>4.0</td>\n",
       "      <td>X</td>\n",
       "      <td>100627108</td>\n",
       "      <td>100639991</td>\n",
       "      <td>-</td>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000003056</th>\n",
       "      <td>CTRL.NEUCV136DHM</td>\n",
       "      <td>4.420148e-12</td>\n",
       "      <td>3.240001e-08</td>\n",
       "      <td>6.50</td>\n",
       "      <td>0.46</td>\n",
       "      <td>3296</td>\n",
       "      <td>2594.60</td>\n",
       "      <td>1884.82</td>\n",
       "      <td>498.90</td>\n",
       "      <td>True</td>\n",
       "      <td>103</td>\n",
       "      <td>1</td>\n",
       "      <td>26.0</td>\n",
       "      <td>12</td>\n",
       "      <td>8940362</td>\n",
       "      <td>8949955</td>\n",
       "      <td>-</td>\n",
       "      <td>M6PR</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000004455</th>\n",
       "      <td>CASE.NEUGW340YEB</td>\n",
       "      <td>4.311222e-12</td>\n",
       "      <td>8.216409e-07</td>\n",
       "      <td>-7.20</td>\n",
       "      <td>-0.75</td>\n",
       "      <td>942</td>\n",
       "      <td>817.30</td>\n",
       "      <td>1374.34</td>\n",
       "      <td>250.69</td>\n",
       "      <td>True</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>33007939</td>\n",
       "      <td>33080996</td>\n",
       "      <td>-</td>\n",
       "      <td>AK2</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000278637</th>\n",
       "      <td>CASE.NEUJA217MTJ</td>\n",
       "      <td>9.435958e-09</td>\n",
       "      <td>1.123952e-04</td>\n",
       "      <td>-6.70</td>\n",
       "      <td>-2.50</td>\n",
       "      <td>120</td>\n",
       "      <td>64.15</td>\n",
       "      <td>367.32</td>\n",
       "      <td>19.69</td>\n",
       "      <td>True</td>\n",
       "      <td>68</td>\n",
       "      <td>1</td>\n",
       "      <td>15.5</td>\n",
       "      <td>6</td>\n",
       "      <td>26021678</td>\n",
       "      <td>26021990</td>\n",
       "      <td>+</td>\n",
       "      <td>HIST1H4A</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000278771</th>\n",
       "      <td>CASE.NEUGW340YEB</td>\n",
       "      <td>4.798981e-10</td>\n",
       "      <td>4.572995e-05</td>\n",
       "      <td>5.05</td>\n",
       "      <td>1.96</td>\n",
       "      <td>4043</td>\n",
       "      <td>1815.72</td>\n",
       "      <td>467.70</td>\n",
       "      <td>12.47</td>\n",
       "      <td>True</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>14</td>\n",
       "      <td>49853615</td>\n",
       "      <td>49853914</td>\n",
       "      <td>-</td>\n",
       "      <td>Metazoa_SRP</td>\n",
       "      <td>misc_RNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000279203</th>\n",
       "      <td>CASE.NEUZP278MR4</td>\n",
       "      <td>2.439758e-09</td>\n",
       "      <td>1.937391e-05</td>\n",
       "      <td>4.83</td>\n",
       "      <td>1.13</td>\n",
       "      <td>268</td>\n",
       "      <td>126.04</td>\n",
       "      <td>57.39</td>\n",
       "      <td>60.20</td>\n",
       "      <td>True</td>\n",
       "      <td>33</td>\n",
       "      <td>1</td>\n",
       "      <td>23.5</td>\n",
       "      <td>19</td>\n",
       "      <td>15383201</td>\n",
       "      <td>15383832</td>\n",
       "      <td>-</td>\n",
       "      <td>AC005785.5</td>\n",
       "      <td>TEC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000279500</th>\n",
       "      <td>CASE.NEUMD869UV7</td>\n",
       "      <td>5.557868e-08</td>\n",
       "      <td>2.648073e-03</td>\n",
       "      <td>-7.39</td>\n",
       "      <td>-6.33</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>93.65</td>\n",
       "      <td>6.88</td>\n",
       "      <td>True</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>4.0</td>\n",
       "      <td>12</td>\n",
       "      <td>128813185</td>\n",
       "      <td>128814750</td>\n",
       "      <td>-</td>\n",
       "      <td>RP11-21K12.2</td>\n",
       "      <td>TEC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000279598</th>\n",
       "      <td>CASE.NEUJA217MTJ</td>\n",
       "      <td>3.441697e-14</td>\n",
       "      <td>2.186417e-09</td>\n",
       "      <td>-9.25</td>\n",
       "      <td>-3.74</td>\n",
       "      <td>14</td>\n",
       "      <td>11.80</td>\n",
       "      <td>169.00</td>\n",
       "      <td>23.57</td>\n",
       "      <td>True</td>\n",
       "      <td>68</td>\n",
       "      <td>1</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2</td>\n",
       "      <td>178554560</td>\n",
       "      <td>178558966</td>\n",
       "      <td>+</td>\n",
       "      <td>RP11-65L3.3</td>\n",
       "      <td>TEC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1358 rows × 19 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                         sampleID        pValue       padjust  zScore  l2fc  \\\n",
       "geneID                                                                        \n",
       "ENSG00000000003  CASE.NEUAA599TMX  7.156284e-18  1.363858e-12   -9.53 -4.51   \n",
       "ENSG00000000003  CTRL.NEUHE723FGT  6.700160e-09  1.276929e-03   -5.45 -2.59   \n",
       "ENSG00000000003  CASE.NEUBA169GXD  4.492763e-08  2.140598e-03   -4.95 -2.36   \n",
       "ENSG00000003056  CTRL.NEUCV136DHM  4.420148e-12  3.240001e-08    6.50  0.46   \n",
       "ENSG00000004455  CASE.NEUGW340YEB  4.311222e-12  8.216409e-07   -7.20 -0.75   \n",
       "...                           ...           ...           ...     ...   ...   \n",
       "ENSG00000278637  CASE.NEUJA217MTJ  9.435958e-09  1.123952e-04   -6.70 -2.50   \n",
       "ENSG00000278771  CASE.NEUGW340YEB  4.798981e-10  4.572995e-05    5.05  1.96   \n",
       "ENSG00000279203  CASE.NEUZP278MR4  2.439758e-09  1.937391e-05    4.83  1.13   \n",
       "ENSG00000279500  CASE.NEUMD869UV7  5.557868e-08  2.648073e-03   -7.39 -6.33   \n",
       "ENSG00000279598  CASE.NEUJA217MTJ  3.441697e-14  2.186417e-09   -9.25 -3.74   \n",
       "\n",
       "                 rawcounts  normcounts  meanCorrected   theta  aberrant  \\\n",
       "geneID                                                                    \n",
       "ENSG00000000003         30       40.08         934.62   19.47      True   \n",
       "ENSG00000000003         77      154.18         934.62   19.47      True   \n",
       "ENSG00000000003        140      181.84         934.62   19.47      True   \n",
       "ENSG00000003056       3296     2594.60        1884.82  498.90      True   \n",
       "ENSG00000004455        942      817.30        1374.34  250.69      True   \n",
       "...                    ...         ...            ...     ...       ...   \n",
       "ENSG00000278637        120       64.15         367.32   19.69      True   \n",
       "ENSG00000278771       4043     1815.72         467.70   12.47      True   \n",
       "ENSG00000279203        268      126.04          57.39   60.20      True   \n",
       "ENSG00000279500          0        0.00          93.65    6.88      True   \n",
       "ENSG00000279598         14       11.80         169.00   23.57      True   \n",
       "\n",
       "                 AberrantBySample  AberrantByGene  padj_rank Chromosome  \\\n",
       "geneID                                                                    \n",
       "ENSG00000000003                 5               3        1.0          X   \n",
       "ENSG00000000003                 5               3        1.0          X   \n",
       "ENSG00000000003                 7               3        4.0          X   \n",
       "ENSG00000003056               103               1       26.0         12   \n",
       "ENSG00000004455                 7               1        1.0          1   \n",
       "...                           ...             ...        ...        ...   \n",
       "ENSG00000278637                68               1       15.5          6   \n",
       "ENSG00000278771                 7               1        2.0         14   \n",
       "ENSG00000279203                33               1       23.5         19   \n",
       "ENSG00000279500                 7               1        4.0         12   \n",
       "ENSG00000279598                68               1        3.0          2   \n",
       "\n",
       "                     Start        End Strand     gene_name    gene_biotype  \n",
       "geneID                                                                      \n",
       "ENSG00000000003  100627108  100639991      -        TSPAN6  protein_coding  \n",
       "ENSG00000000003  100627108  100639991      -        TSPAN6  protein_coding  \n",
       "ENSG00000000003  100627108  100639991      -        TSPAN6  protein_coding  \n",
       "ENSG00000003056    8940362    8949955      -          M6PR  protein_coding  \n",
       "ENSG00000004455   33007939   33080996      -           AK2  protein_coding  \n",
       "...                    ...        ...    ...           ...             ...  \n",
       "ENSG00000278637   26021678   26021990      +      HIST1H4A  protein_coding  \n",
       "ENSG00000278771   49853615   49853914      -   Metazoa_SRP        misc_RNA  \n",
       "ENSG00000279203   15383201   15383832      -    AC005785.5             TEC  \n",
       "ENSG00000279500  128813185  128814750      -  RP11-21K12.2             TEC  \n",
       "ENSG00000279598  178554560  178558966      +   RP11-65L3.3             TEC  \n",
       "\n",
       "[1358 rows x 19 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_results = pd.read_csv(snakemake.input['results']).set_index('geneID')\n",
    "df_genes = pd.read_csv(snakemake.input['genes']).set_index('geneID')\n",
    "df_results = df_results.join(df_genes)\n",
    "df_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "40c52eef-886d-4e9c-b444-dcc7c762d38a",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_results = df_results[df_results['Chromosome'] == snakemake.wildcards['chrom'].replace('chr', '')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "c928d3f1-f7d2-4f50-b607-26cbe63f84d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_parquet(snakemake.input['vep'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "cf389252-8ee8-4a52-b04c-6e808227086b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1780776, 89)\n",
      "(4498, 89)\n",
      "(151, 89)\n"
     ]
    }
   ],
   "source": [
    "df = df[df['AF'].fillna(0) < 0.0001] # subset outliers\n",
    "print(df.shape)\n",
    "df = df[df['IMPACT'].isin({'HIGH', 'MODERATE'})] # subset for high impact variants\n",
    "print(df.shape)\n",
    "df = df[df['Gene'].isin(set(df_results.index))] # subset outliers\n",
    "print(df.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "5c252069-69cb-4ec0-9fcd-2d9c57021da0",
   "metadata": {},
   "outputs": [],
   "source": [
    "vcf = MultiSampleVCF(snakemake.input['vcf'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "942a136a-b30d-4d13-9ad0-c7c96244fc5d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Participant_ID</th>\n",
       "      <th>omic</th>\n",
       "      <th>CGND_ID</th>\n",
       "      <th>iPSC_Line</th>\n",
       "      <th>line</th>\n",
       "      <th>differentiation</th>\n",
       "      <th>data_level</th>\n",
       "      <th>experiment</th>\n",
       "      <th>path</th>\n",
       "      <th>size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CASE-NEUAA599TMX</td>\n",
       "      <td>genomics</td>\n",
       "      <td>CGND-HDA-00461</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>WGS</td>\n",
       "      <td>genomics/2_bam/CASE-NEUAA599TMX/CASE-NEUAA599T...</td>\n",
       "      <td>23079388468</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CASE-NEUAA599TMX</td>\n",
       "      <td>genomics</td>\n",
       "      <td>CGND-HDA-00461</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>WGS</td>\n",
       "      <td>genomics/2_bam/CASE-NEUAA599TMX/CASE-NEUAA599T...</td>\n",
       "      <td>1582288</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CASE-NEUAB000NKC</td>\n",
       "      <td>genomics</td>\n",
       "      <td>CGND-HDA-00583</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>WGS</td>\n",
       "      <td>genomics/2_bam/CASE-NEUAB000NKC/CASE-NEUAB000N...</td>\n",
       "      <td>22547466160</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CASE-NEUAB000NKC</td>\n",
       "      <td>genomics</td>\n",
       "      <td>CGND-HDA-00583</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>WGS</td>\n",
       "      <td>genomics/2_bam/CASE-NEUAB000NKC/CASE-NEUAB000N...</td>\n",
       "      <td>1684005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CASE-NEUAE993EPR</td>\n",
       "      <td>genomics</td>\n",
       "      <td>CGND-HDA-03249</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>WGS</td>\n",
       "      <td>genomics/2_bam/CASE-NEUAE993EPR/CASE-NEUAE993E...</td>\n",
       "      <td>15054503804</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4777</th>\n",
       "      <td>CTRL-NEURJ861MMD</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>proteomics/4_matrix/ProteinLevel_v4-1_release_...</td>\n",
       "      <td>5143400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4778</th>\n",
       "      <td>CTRL-NEUUV825HYF</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>proteomics/4_matrix/ProteinLevel_v4-1_release_...</td>\n",
       "      <td>5143400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4779</th>\n",
       "      <td>CTRL-NEUVZ050YX7</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>proteomics/4_matrix/ProteinLevel_v4-1_release_...</td>\n",
       "      <td>5143400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4780</th>\n",
       "      <td>CTRL-NEUWT164JRQ</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>proteomics/4_matrix/ProteinLevel_v4-1_release_...</td>\n",
       "      <td>5143400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4781</th>\n",
       "      <td>CTRL-NEUYM205MRL</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>proteomics/4_matrix/ProteinLevel_v4-1_release_...</td>\n",
       "      <td>5143400</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4782 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Participant_ID        omic         CGND_ID iPSC_Line line  \\\n",
       "0     CASE-NEUAA599TMX    genomics  CGND-HDA-00461       NaN  NaN   \n",
       "1     CASE-NEUAA599TMX    genomics  CGND-HDA-00461       NaN  NaN   \n",
       "2     CASE-NEUAB000NKC    genomics  CGND-HDA-00583       NaN  NaN   \n",
       "3     CASE-NEUAB000NKC    genomics  CGND-HDA-00583       NaN  NaN   \n",
       "4     CASE-NEUAE993EPR    genomics  CGND-HDA-03249       NaN  NaN   \n",
       "...                ...         ...             ...       ...  ...   \n",
       "4777  CTRL-NEURJ861MMD  proteomics             NaN       NaN  NaN   \n",
       "4778  CTRL-NEUUV825HYF  proteomics             NaN       NaN  NaN   \n",
       "4779  CTRL-NEUVZ050YX7  proteomics             NaN       NaN  NaN   \n",
       "4780  CTRL-NEUWT164JRQ  proteomics             NaN       NaN  NaN   \n",
       "4781  CTRL-NEUYM205MRL  proteomics             NaN       NaN  NaN   \n",
       "\n",
       "     differentiation  data_level experiment  \\\n",
       "0                NaN           2        WGS   \n",
       "1                NaN           2        WGS   \n",
       "2                NaN           2        WGS   \n",
       "3                NaN           2        WGS   \n",
       "4                NaN           2        WGS   \n",
       "...              ...         ...        ...   \n",
       "4777             NaN           4        NaN   \n",
       "4778             NaN           4        NaN   \n",
       "4779             NaN           4        NaN   \n",
       "4780             NaN           4        NaN   \n",
       "4781             NaN           4        NaN   \n",
       "\n",
       "                                                   path         size  \n",
       "0     genomics/2_bam/CASE-NEUAA599TMX/CASE-NEUAA599T...  23079388468  \n",
       "1     genomics/2_bam/CASE-NEUAA599TMX/CASE-NEUAA599T...      1582288  \n",
       "2     genomics/2_bam/CASE-NEUAB000NKC/CASE-NEUAB000N...  22547466160  \n",
       "3     genomics/2_bam/CASE-NEUAB000NKC/CASE-NEUAB000N...      1684005  \n",
       "4     genomics/2_bam/CASE-NEUAE993EPR/CASE-NEUAE993E...  15054503804  \n",
       "...                                                 ...          ...  \n",
       "4777  proteomics/4_matrix/ProteinLevel_v4-1_release_...      5143400  \n",
       "4778  proteomics/4_matrix/ProteinLevel_v4-1_release_...      5143400  \n",
       "4779  proteomics/4_matrix/ProteinLevel_v4-1_release_...      5143400  \n",
       "4780  proteomics/4_matrix/ProteinLevel_v4-1_release_...      5143400  \n",
       "4781  proteomics/4_matrix/ProteinLevel_v4-1_release_...      5143400  \n",
       "\n",
       "[4782 rows x 10 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "1f3c7407-981a-4427-86a8-8203b31c62f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_mapping = {\n",
    "    row['CGND_ID']: row['Participant_ID']\n",
    "    for i, row in df_metadata[df_metadata['omic'] == 'genomics'].iterrows()\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "2773be56-9f66-48df-bcd2-5157eb76f5c4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Variant(chrom='chr1', pos=16378, ref='T', alt='C', id='None',...)"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "next(vcf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "03f94ad8-f47d-4e8b-b438-d3fe897b15fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "df['variant'] = snakemake.wildcards['chrom'] + ':' + df['pos'].astype(str) + ':'+ df['ref'] + '>' + df['alt']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "e79342d9-123f-4aee-b2a8-7272dcf9719a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Feature_type</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>cDNA_position</th>\n",
       "      <th>CDS_position</th>\n",
       "      <th>Protein_position</th>\n",
       "      <th>...</th>\n",
       "      <th>miRNA83</th>\n",
       "      <th>CADD_PHRED</th>\n",
       "      <th>CADD_RAW</th>\n",
       "      <th>condel_prediction</th>\n",
       "      <th>condel_score</th>\n",
       "      <th>sift_prediction</th>\n",
       "      <th>sift_score</th>\n",
       "      <th>polyphen_prediction</th>\n",
       "      <th>polyphen_score</th>\n",
       "      <th>variant</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>15992</th>\n",
       "      <td>5768222</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSG00000057608</td>\n",
       "      <td>ENST00000380181</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>1000/1408</td>\n",
       "      <td>847/1203</td>\n",
       "      <td>283/400</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>42.000000</td>\n",
       "      <td>8.330135</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr10:5768222:G&gt;A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15993</th>\n",
       "      <td>5768222</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSG00000057608</td>\n",
       "      <td>ENST00000380191</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>1158/2297</td>\n",
       "      <td>982/1338</td>\n",
       "      <td>328/445</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>42.000000</td>\n",
       "      <td>8.330135</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr10:5768222:G&gt;A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15994</th>\n",
       "      <td>5768222</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSG00000057608</td>\n",
       "      <td>ENST00000447751</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>466/834</td>\n",
       "      <td>466/612</td>\n",
       "      <td>156/203</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>42.000000</td>\n",
       "      <td>8.330135</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr10:5768222:G&gt;A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16310</th>\n",
       "      <td>5773889</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>ENSG00000057608</td>\n",
       "      <td>ENST00000380181</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>790/1408</td>\n",
       "      <td>637/1203</td>\n",
       "      <td>213/400</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>19.770000</td>\n",
       "      <td>2.067882</td>\n",
       "      <td>neutral</td>\n",
       "      <td>0.005</td>\n",
       "      <td>tolerated</td>\n",
       "      <td>0.63</td>\n",
       "      <td>benign</td>\n",
       "      <td>0.031</td>\n",
       "      <td>chr10:5773889:T&gt;C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16311</th>\n",
       "      <td>5773889</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>ENSG00000057608</td>\n",
       "      <td>ENST00000380191</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>948/2297</td>\n",
       "      <td>772/1338</td>\n",
       "      <td>258/445</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>19.770000</td>\n",
       "      <td>2.067882</td>\n",
       "      <td>neutral</td>\n",
       "      <td>0.006</td>\n",
       "      <td>tolerated</td>\n",
       "      <td>0.63</td>\n",
       "      <td>benign</td>\n",
       "      <td>0.043</td>\n",
       "      <td>chr10:5773889:T&gt;C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3867207</th>\n",
       "      <td>124405471</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>ENSG00000065154</td>\n",
       "      <td>ENST00000368845</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>693/2039</td>\n",
       "      <td>613/1320</td>\n",
       "      <td>205/439</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>19.059999</td>\n",
       "      <td>1.974205</td>\n",
       "      <td>neutral</td>\n",
       "      <td>0.029</td>\n",
       "      <td>tolerated</td>\n",
       "      <td>0.29</td>\n",
       "      <td>benign</td>\n",
       "      <td>0.003</td>\n",
       "      <td>chr10:124405471:C&gt;T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3867214</th>\n",
       "      <td>124405471</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>ENSG00000065154</td>\n",
       "      <td>ENST00000539214</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>518/1864</td>\n",
       "      <td>199/906</td>\n",
       "      <td>67/301</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>19.059999</td>\n",
       "      <td>1.974205</td>\n",
       "      <td>neutral</td>\n",
       "      <td>0.027</td>\n",
       "      <td>tolerated</td>\n",
       "      <td>0.31</td>\n",
       "      <td>benign</td>\n",
       "      <td>0.003</td>\n",
       "      <td>chr10:124405471:C&gt;T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3868536</th>\n",
       "      <td>125839128</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>ENSG00000089876</td>\n",
       "      <td>ENST00000284690</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>2222/3050</td>\n",
       "      <td>1754/2232</td>\n",
       "      <td>585/743</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>21.799999</td>\n",
       "      <td>2.337842</td>\n",
       "      <td>neutral</td>\n",
       "      <td>0.300</td>\n",
       "      <td>tolerated</td>\n",
       "      <td>0.09</td>\n",
       "      <td>benign</td>\n",
       "      <td>0.054</td>\n",
       "      <td>chr10:125839128:T&gt;G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3868537</th>\n",
       "      <td>125839128</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>ENSG00000089876</td>\n",
       "      <td>ENST00000368721</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>1561/2389</td>\n",
       "      <td>626/1104</td>\n",
       "      <td>209/367</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>21.799999</td>\n",
       "      <td>2.337842</td>\n",
       "      <td>neutral</td>\n",
       "      <td>0.349</td>\n",
       "      <td>deleterious</td>\n",
       "      <td>0.05</td>\n",
       "      <td>benign</td>\n",
       "      <td>0.054</td>\n",
       "      <td>chr10:125839128:T&gt;G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4023649</th>\n",
       "      <td>131311321</td>\n",
       "      <td>C</td>\n",
       "      <td>CGCG</td>\n",
       "      <td>ENSG00000176769</td>\n",
       "      <td>ENST00000368642</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>inframe_insertion</td>\n",
       "      <td>400-401/2618</td>\n",
       "      <td>314-315/1761</td>\n",
       "      <td>105/586</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>16.110001</td>\n",
       "      <td>1.558560</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr10:131311321:C&gt;CGCG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>151 rows × 90 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               pos ref   alt             Gene          Feature Feature_type  \\\n",
       "15992      5768222   G     A  ENSG00000057608  ENST00000380181   Transcript   \n",
       "15993      5768222   G     A  ENSG00000057608  ENST00000380191   Transcript   \n",
       "15994      5768222   G     A  ENSG00000057608  ENST00000447751   Transcript   \n",
       "16310      5773889   T     C  ENSG00000057608  ENST00000380181   Transcript   \n",
       "16311      5773889   T     C  ENSG00000057608  ENST00000380191   Transcript   \n",
       "...            ...  ..   ...              ...              ...          ...   \n",
       "3867207  124405471   C     T  ENSG00000065154  ENST00000368845   Transcript   \n",
       "3867214  124405471   C     T  ENSG00000065154  ENST00000539214   Transcript   \n",
       "3868536  125839128   T     G  ENSG00000089876  ENST00000284690   Transcript   \n",
       "3868537  125839128   T     G  ENSG00000089876  ENST00000368721   Transcript   \n",
       "4023649  131311321   C  CGCG  ENSG00000176769  ENST00000368642   Transcript   \n",
       "\n",
       "               Consequence cDNA_position  CDS_position Protein_position  ...  \\\n",
       "15992          stop_gained     1000/1408      847/1203          283/400  ...   \n",
       "15993          stop_gained     1158/2297      982/1338          328/445  ...   \n",
       "15994          stop_gained       466/834       466/612          156/203  ...   \n",
       "16310     missense_variant      790/1408      637/1203          213/400  ...   \n",
       "16311     missense_variant      948/2297      772/1338          258/445  ...   \n",
       "...                    ...           ...           ...              ...  ...   \n",
       "3867207   missense_variant      693/2039      613/1320          205/439  ...   \n",
       "3867214   missense_variant      518/1864       199/906           67/301  ...   \n",
       "3868536   missense_variant     2222/3050     1754/2232          585/743  ...   \n",
       "3868537   missense_variant     1561/2389      626/1104          209/367  ...   \n",
       "4023649  inframe_insertion  400-401/2618  314-315/1761          105/586  ...   \n",
       "\n",
       "        miRNA83 CADD_PHRED  CADD_RAW  condel_prediction condel_score  \\\n",
       "15992      None  42.000000  8.330135               None          NaN   \n",
       "15993      None  42.000000  8.330135               None          NaN   \n",
       "15994      None  42.000000  8.330135               None          NaN   \n",
       "16310      None  19.770000  2.067882            neutral        0.005   \n",
       "16311      None  19.770000  2.067882            neutral        0.006   \n",
       "...         ...        ...       ...                ...          ...   \n",
       "3867207    None  19.059999  1.974205            neutral        0.029   \n",
       "3867214    None  19.059999  1.974205            neutral        0.027   \n",
       "3868536    None  21.799999  2.337842            neutral        0.300   \n",
       "3868537    None  21.799999  2.337842            neutral        0.349   \n",
       "4023649    None  16.110001  1.558560               None          NaN   \n",
       "\n",
       "         sift_prediction  sift_score polyphen_prediction polyphen_score  \\\n",
       "15992               None         NaN                None            NaN   \n",
       "15993               None         NaN                None            NaN   \n",
       "15994               None         NaN                None            NaN   \n",
       "16310          tolerated        0.63              benign          0.031   \n",
       "16311          tolerated        0.63              benign          0.043   \n",
       "...                  ...         ...                 ...            ...   \n",
       "3867207        tolerated        0.29              benign          0.003   \n",
       "3867214        tolerated        0.31              benign          0.003   \n",
       "3868536        tolerated        0.09              benign          0.054   \n",
       "3868537      deleterious        0.05              benign          0.054   \n",
       "4023649             None         NaN                None            NaN   \n",
       "\n",
       "                        variant  \n",
       "15992         chr10:5768222:G>A  \n",
       "15993         chr10:5768222:G>A  \n",
       "15994         chr10:5768222:G>A  \n",
       "16310         chr10:5773889:T>C  \n",
       "16311         chr10:5773889:T>C  \n",
       "...                         ...  \n",
       "3867207     chr10:124405471:C>T  \n",
       "3867214     chr10:124405471:C>T  \n",
       "3868536     chr10:125839128:T>G  \n",
       "3868537     chr10:125839128:T>G  \n",
       "4023649  chr10:131311321:C>CGCG  \n",
       "\n",
       "[151 rows x 90 columns]"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "add61c45-5951-460f-85e4-a75ed8d28fb6",
   "metadata": {},
   "outputs": [],
   "source": [
    "df['samples'] = [\n",
    "    [\n",
    "        sample_mapping[k.replace('-b38', '')]\n",
    "        for k, v in vcf.get_samples(vcf.get_variant(i)).items()\n",
    "    ]\n",
    "    for i in df['variant']\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "417e80ff-bb01-4da1-bef3-de62e74f2bf7",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df[df['samples'].map(len) < 7].explode('samples')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "d62980cd-6d5c-4302-8c37-34f893e050f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_csv(snakemake.output['prioritized'], index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:anaconda-als]",
   "language": "python",
   "name": "conda-env-anaconda-als-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
